core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/ConsistencyTracker.java

```diff
@@ -0,0 +1,54 @@
+package ai.timefold.solver.core.impl.domain.variable.declarative;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+
+import ai.timefold.solver.core.impl.domain.entity.descriptor.EntityDescriptor;
+import ai.timefold.solver.core.impl.domain.solution.descriptor.SolutionDescriptor;
+
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class ConsistencyTracker<Solution_> {
+    private final IdentityHashMap<Object, Boolean> entityToIsInconsistentMap = new IdentityHashMap<>();
+    private final HashMap<Class<?>, EntityConsistencyState<Solution_, Object>> entityClassToConsistencyStateMap =
+            new HashMap<>();
+    private boolean isFrozen;
+
+    public EntityConsistencyState<Solution_, Object>
+            getDeclarativeEntityConsistencyState(EntityDescriptor<Solution_> entityDescriptor) {
+        return entityClassToConsistencyStateMap.computeIfAbsent(entityDescriptor.getEntityClass(),
+                ignored -> new EntityConsistencyState<>(entityDescriptor, entityToIsInconsistentMap));
+    }
+
+    public void setUnknownConsistencyFromEntityShadowVariablesInconsistent(SolutionDescriptor<Solution_> solutionDescriptor,
```

Comment:
The concept of "unknown inconsistency" ought to be explained in Javadoc. Either here, or on `Graph.setUnknownInconsistencyValues()`.